### PR TITLE
fix: Mermaid and Math rendering in Jupyter Lab and VSCode

### PR DESCRIPTION
**Pull request by iisakkirotko**
Should fix https://github.com/widgetti/solara/issues/175, and partially fix https://github.com/widgetti/solara/issues/446

### CODE REVIEW

**Comment by iisakkirotko** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/widgetti/solara/480?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #479** <a href="https://app.graphite.dev/github/pr/widgetti/solara/479?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #480** <a href="https://app.graphite.dev/github/pr/widgetti/solara/480?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/widgetti/solara/480?utm_source=stack-comment).

File: packages/solara-widget-manager/package.json [hunk 1]
```diff
@@ -28,9 +28,11 @@
     "@lumino/signaling": "^1.4.3",
     "@lumino/virtualdom": "^1.8.0",
     "@lumino/widgets": "^1.18.0",
-    "mathjax-full": "^3.0.0"
+    "mathjax-full": "^3.0.0",
+    "mermaid": "^8.6.4"
```
Comment (maartenbreddels) [inline packages/solara-widget-manager/package.json:32]:
```suggestion
```

File: packages/solara-widget-manager/src/mermaid.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import mermaid from 'mermaid';
+
+mermaid.initialize({
+  startOnLoad: true,
+});
+
+export function renderMermaid(): void {
+  mermaid.init();
+}
```
Comment (maartenbreddels) [inline packages/solara-widget-manager/src/mermaid.ts:9]:
```suggestion
```

File: packages/solara-widget-manager8/package.json [hunk 3]
```diff
@@ -26,9 +26,11 @@
     "@lumino/virtualdom": "^1.8.0",
     "@lumino/widgets": "^1.18.0",
     "mathjax-full": "^3.0.0",
+    "mermaid": "^8.6.4",
```
Comment (maartenbreddels) [inline packages/solara-widget-manager8/package.json:29]:
```suggestion
```

File: solara/components/markdown.py [hunk 4]
```diff
@@ -109,26 +134,114 @@ def _markdown_template(html, style=""):
...
+            if (!base.endsWith('/')) {
+                base += '/';
+            }
+            return base
+        },
+        getCdn() {
+    """
+        + (
+            """
+            return "https://cdn.jsdelivr.net/npm";
```
Comment (maartenbreddels) [inline solara/components/markdown.py:215]:
```suggestion
            return "{solara.server.settings.assets.cdn!r}";
```

File: solara/components/markdown.py [hunk 5]
```diff
@@ -109,26 +134,114 @@ def _markdown_template(html, style=""):
...
+        }
+    },
+        """
+            if vscode
```
Comment (maartenbreddels) [inline solara/components/markdown.py:219]:
```suggestion
            if solara.server.settings.assets.proxy
```

File: solara/components/markdown.py [hunk 6]
```diff
@@ -276,9 +389,14 @@ def highlight(src, language, *args, **kwargs):
...
+
+    VSCODE_PID = os.getenv("VSCODE_PID")
+
     md = solara.use_memo(make_markdown_object, dependencies=[unsafe_solara_execute])
     html = md.convert(md_text)
     # if we update the template value, the whole vue tree will rerender (ipvue/ipyvuetify issue)
     # however, using the hash we simply generate a new widget each time
     hash = hashlib.sha256((html + str(unsafe_solara_execute)).encode("utf-8")).hexdigest()
-    return v.VuetifyTemplate.element(template=_markdown_template(html, style)).key(hash)
+    return v.VuetifyTemplate.element(template=_markdown_template(html, style, vscode=VSCODE_PID is not None)).key(hash)
```
Comment (maartenbreddels) [inline solara/components/markdown.py:402]:
```suggestion
    return v.VuetifyTemplate.element(template=_markdown_template(html, style).key(hash)
```

File: solara/components/markdown.py [hunk 7]
```diff
@@ -66,24 +66,51 @@ def _run_solara(code):
     )
 
 
-def _markdown_template(html, style=""):
-    return (
+def _markdown_template(html, style="", vscode=False):
```
Comment (maartenbreddels) [inline solara/components/markdown.py:69]:
```suggestion
def _markdown_template(html, style="",):
```

**Review by maartenbreddels — COMMENTED** [review]
Lets go with the changes we discussed over a call, and keep the katex part in

**Comment by maartenbreddels** [comment]
See https://github.com/widgetti/solara/pull/487 and https://github.com/widgetti/solara/pull/488 for the refactor based on our discussions yesterday.

File: packages/solara-vuetify-app/src/solara.js [hunk 8]
```diff
@@ -1 +1 @@
-export { RenderMimeRegistry, WidgetManager, connectKernel, extendedRendererFactories, renderMathJax, shutdownKernel } from '@widgetti/solara-widget-manager';
+export { RenderMimeRegistry, WidgetManager, connectKernel, extendedRendererFactories, renderMathJax, renderMermaid, shutdownKernel } from '@widgetti/solara-widget-manager';
```
Comment (maartenbreddels) [inline packages/solara-vuetify-app/src/solara.js:1]: renderMermaid shouldn't be there anymore right?

File: packages/solara-vuetify3-app/src/solara.js [hunk 9]
```diff
@@ -1 +1 @@
-export { RenderMimeRegistry, WidgetManager, connectKernel, extendedRendererFactories, renderMathJax, shutdownKernel } from '@widgetti/solara-widget-manager';
+export { RenderMimeRegistry, WidgetManager, connectKernel, extendedRendererFactories, renderMathJax, renderMermaid, shutdownKernel } from '@widgetti/solara-widget-manager';
```
Comment (maartenbreddels) [inline packages/solara-vuetify3-app/src/solara.js:1]: same here

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
After a small change we can merge I think. Surprised it does not fail the build on the renderMermaid actually!

**Pull request merged**